### ISSUE TITLE
Generating an activity stream from a wiki-like system: 

### ISSUE THREAD

**Issue by wirehead** [jasnell/w3c-socialwg-activitystreams#237]
I'm working on a fun-hack project that works a lot like a wiki and I'm adding activity streams support to it, which means that the internal query functions are returning ActivityStreams (currently a grunting caveperson version thereof, but I'm going to migrate it to 2.0) and then the display function is working against the ActivityStream.

The underlying data model is that users edit pages, of which the most current rev is always available from the base URL, but previous revisions are available available via a query string param.

There's some contention there, because there's a very pervasive context in such a system of the "head" revision that isn't there for most social activities.

Right now the `url` property in the `Object` for an activity points to the base URL for the head revision and the fully-disambiguated url with the revision ID in the query string for all non-head revisions.  Similarly the `id` in the `Object` object just includes the underlying entity ID independent of revision, but the `id` in the `Activity` object does include the revision ID.

I'm not sure if this is the right way to do things, but it seems like a fairly naive version of the Activity Streams parser will generally do the right thing.  Should I instead be linking directly to the head revision all the time or directly to the fully-disambiguated url all of the time?  If so, how would I allow a general-purpose Activity Streams display routine be able to disambiguate revisions?

**Comment by jasnell** [comment]
Could you provide an example of the serialization?

**Comment by wirehead** [comment]
Sure, so here's an activity stream for the root page of a test site, still in caveperson-grunting 1.0, not 2.0:

```javascript
{
  "items": [
    {
      "object": {
        "url": "/?revisionId=2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483",
        "id": "urn:wh"
      },
      "verb": "post",
      "updated": "2015-09-13T22:17:11.661Z",
      "startTime": "2015-09-13T22:17:11.661Z",
      "endTime": "2015-09-13T22:17:11.662Z",
      "rm3:revisionNum": 1,
      "rm3:revisionId": "2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483",
      "id": "urn:2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483:1",
      "published": "2015-09-13T22:17:11.662Z",
      "actor": {
        "objectType": "site",
        "id": "urn:root"
      }
    },
    {
      "object": {
        "url": "/?revisionId=96e94050-6561-11e5-81a6-afdb68195b27",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-09-27T21:49:12.277Z",
      "startTime": "2015-09-27T21:49:12.277Z",
      "endTime": "2015-09-27T21:49:12.280Z",
      "rm3:revisionNum": 2,
      "rm3:revisionId": "96e94050-6561-11e5-81a6-afdb68195b27",
      "id": "urn:96e94050-6561-11e5-81a6-afdb68195b27:2",
      "published": "2015-09-27T21:49:12.280Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "object": {
        "url": "/?revisionId=acea7e30-7058-11e5-9eba-f3666a003ebf",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-10-11T20:43:06.515Z",
      "startTime": "2015-10-11T20:43:06.515Z",
      "endTime": "2015-10-11T20:43:06.526Z",
      "rm3:revisionNum": 3,
      "rm3:revisionId": "acea7e30-7058-11e5-9eba-f3666a003ebf",
      "id": "urn:acea7e30-7058-11e5-9eba-f3666a003ebf:acea7e30-7058-11e5-9eba-f3666a003ebf",
      "published": "2015-10-11T20:43:06.526Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "object": {
        "url": "/?revisionId=d9914f40-7058-11e5-8a77-691e032e5388",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-10-11T20:44:21.428Z",
      "startTime": "2015-10-11T20:44:21.428Z",
      "endTime": "2015-10-11T20:44:21.431Z",
      "rm3:revisionNum": 4,
      "rm3:revisionId": "d9914f40-7058-11e5-8a77-691e032e5388",
      "id": "urn:d9914f40-7058-11e5-8a77-691e032e5388:d9914f40-7058-11e5-8a77-691e032e5388",
      "published": "2015-10-11T20:44:21.431Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "object": {
        "url": "/?revisionId=388e2c70-7059-11e5-bb3d-cd70f11da9e4",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-10-11T20:47:00.791Z",
      "startTime": "2015-10-11T20:47:00.791Z",
      "endTime": "2015-10-11T20:47:00.802Z",
      "rm3:revisionNum": 5,
      "rm3:revisionId": "388e2c70-7059-11e5-bb3d-cd70f11da9e4",
      "id": "urn:388e2c70-7059-11e5-bb3d-cd70f11da9e4:388e2c70-7059-11e5-bb3d-cd70f11da9e4",
      "published": "2015-10-11T20:47:00.802Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "object": {
        "url": "/?revisionId=d4fe3d20-7059-11e5-8f5b-8536e856d478",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-10-11T20:51:23.250Z",
      "startTime": "2015-10-11T20:51:23.250Z",
      "endTime": "2015-10-11T20:51:23.280Z",
      "rm3:revisionNum": 6,
      "rm3:revisionId": "d4fe3d20-7059-11e5-8f5b-8536e856d478",
      "id": "urn:d4fe3d20-7059-11e5-8f5b-8536e856d478:d4fe3d20-7059-11e5-8f5b-8536e856d478",
      "published": "2015-10-11T20:51:23.280Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "object": {
        "url": "/",
        "id": "urn:wh"
      },
      "verb": "update",
      "updated": "2015-10-11T20:55:49.135Z",
      "startTime": "2015-10-11T20:55:49.135Z",
      "endTime": "2015-10-11T20:55:49.140Z",
      "rm3:revisionNum": 7,
      "rm3:revisionId": "737911f0-705a-11e5-91c7-15dfd36cd712",
      "id": "urn:737911f0-705a-11e5-91c7-15dfd36cd712:737911f0-705a-11e5-91c7-15dfd36cd712",
      "published": "2015-10-11T20:55:49.140Z",
      "actor": {
        "objectType": "person",
        "id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    }
  ]
}
```

**Comment by jasnell** [comment]
Here's the equivalent according to the current AS2 Working Draft. Overall this looks just fine. You might consider looking that the `context` property as means of relating each activity back to a common root. 

(Note: it's quite likely that AS2 will be updated so that `@id` and `@type` are aliased to `id` and `type`)

```json
{
  "@context": ["http://www.w3.org/ns/activitystreams", {
    "rm3": "http://rm3.example.org/#"
  }],
  "items": [
    {
      "@type": "Create",
      "object": {
        "url": "/?revisionId=2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483",
        "@id": "urn:wh"
      },
      "updated": "2015-09-13T22:17:11.661Z",
      "startTime": "2015-09-13T22:17:11.661Z",
      "endTime": "2015-09-13T22:17:11.662Z",
      "rm3:revisionNum": 1,
      "rm3:revisionId": "2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483",
      "@id": "urn:2e1e90e0-5a65-11e5-a6a9-b7f6dbb22483:1",
      "published": "2015-09-13T22:17:11.662Z",
      "actor": {
        "@type": "http://example.org/Site",
        "@id": "urn:root"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/?revisionId=96e94050-6561-11e5-81a6-afdb68195b27",
        "@id": "urn:wh"
      },
      "updated": "2015-09-27T21:49:12.277Z",
      "startTime": "2015-09-27T21:49:12.277Z",
      "endTime": "2015-09-27T21:49:12.280Z",
      "rm3:revisionNum": 2,
      "rm3:revisionId": "96e94050-6561-11e5-81a6-afdb68195b27",
      "@id": "urn:96e94050-6561-11e5-81a6-afdb68195b27:2",
      "published": "2015-09-27T21:49:12.280Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/?revisionId=acea7e30-7058-11e5-9eba-f3666a003ebf",
        "@id": "urn:wh"
      },
      "updated": "2015-10-11T20:43:06.515Z",
      "startTime": "2015-10-11T20:43:06.515Z",
      "endTime": "2015-10-11T20:43:06.526Z",
      "rm3:revisionNum": 3,
      "rm3:revisionId": "acea7e30-7058-11e5-9eba-f3666a003ebf",
      "@id": "urn:acea7e30-7058-11e5-9eba-f3666a003ebf:acea7e30-7058-11e5-9eba-f3666a003ebf",
      "published": "2015-10-11T20:43:06.526Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/?revisionId=d9914f40-7058-11e5-8a77-691e032e5388",
        "@id": "urn:wh"
      },
      "updated": "2015-10-11T20:44:21.428Z",
      "startTime": "2015-10-11T20:44:21.428Z",
      "endTime": "2015-10-11T20:44:21.431Z",
      "rm3:revisionNum": 4,
      "rm3:revisionId": "d9914f40-7058-11e5-8a77-691e032e5388",
      "@id": "urn:d9914f40-7058-11e5-8a77-691e032e5388:d9914f40-7058-11e5-8a77-691e032e5388",
      "published": "2015-10-11T20:44:21.431Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/?revisionId=388e2c70-7059-11e5-bb3d-cd70f11da9e4",
        "@id": "urn:wh"
      },
      "updated": "2015-10-11T20:47:00.791Z",
      "startTime": "2015-10-11T20:47:00.791Z",
      "endTime": "2015-10-11T20:47:00.802Z",
      "rm3:revisionNum": 5,
      "rm3:revisionId": "388e2c70-7059-11e5-bb3d-cd70f11da9e4",
      "@id": "urn:388e2c70-7059-11e5-bb3d-cd70f11da9e4:388e2c70-7059-11e5-bb3d-cd70f11da9e4",
      "published": "2015-10-11T20:47:00.802Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/?revisionId=d4fe3d20-7059-11e5-8f5b-8536e856d478",
        "@id": "urn:wh"
      },
      "updated": "2015-10-11T20:51:23.250Z",
      "startTime": "2015-10-11T20:51:23.250Z",
      "endTime": "2015-10-11T20:51:23.280Z",
      "rm3:revisionNum": 6,
      "rm3:revisionId": "d4fe3d20-7059-11e5-8f5b-8536e856d478",
      "@id": "urn:d4fe3d20-7059-11e5-8f5b-8536e856d478:d4fe3d20-7059-11e5-8f5b-8536e856d478",
      "published": "2015-10-11T20:51:23.280Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    },
    {
      "@type": "Update",
      "object": {
        "url": "/",
        "@id": "urn:wh"
      },
      "updated": "2015-10-11T20:55:49.135Z",
      "startTime": "2015-10-11T20:55:49.135Z",
      "endTime": "2015-10-11T20:55:49.140Z",
      "rm3:revisionNum": 7,
      "rm3:revisionId": "737911f0-705a-11e5-91c7-15dfd36cd712",
      "@id": "urn:737911f0-705a-11e5-91c7-15dfd36cd712:737911f0-705a-11e5-91c7-15dfd36cd712",
      "published": "2015-10-11T20:55:49.140Z",
      "actor": {
        "@type": "Person",
        "@id": "urn:wh.users.wirehead",
        "url": "/users/wirehead/"
      }
    }
  ]
}
```

**Comment by jasnell** [comment]
@wirehead ... I'm going to go ahead and close this issue but feel free to continue the conversation here or on the mailing list or IRC channels.
